### PR TITLE
Implement Proof of Work

### PR DESCRIPTION
**Pull request by emmorais**
Implementation of Proof of Work:
* TOTAL_SIZE=14 was chosen  so that it takes roughly 20 seconds in my laptop. We can calibrate it to increase the difficulty if necessary.

### CODE REVIEW

File: setup1-contributor/src/reliability/cpu.rs [hunk 1]
```diff
@@ -24,3 +57,25 @@ where
...
+
+    #[test]
+    fn test_calculate_powers_of_tau() {
+        let data: Vec<u8> = vec![0; DATA_LENGTH];
+        let now = Instant::now();
+        let output = calculate_powers_of_tau::<Bls12_377>(data);
+        let elapsed = now.elapsed();
+        println!("Proof of Work took: {:.2?}", elapsed);
+        assert_eq!(output.len(), OUTPUT_LENGTH);
+        let expected = [12, 53, 232, 21, 121, 10, 125, 129];
```
Comment (ibaryshnikov) [inline setup1-contributor/src/reliability/cpu.rs:78]: am I missing something or shouldn't 0 power n still be 0?

File: setup1-contributor/src/reliability/cpu.rs [hunk 2]
```diff
@@ -2,20 +2,53 @@ use std::fmt::Debug;
...
+    Phase1::initialization(&mut output, UseCompression::Yes, &parameters).unwrap();
+    let mut input = vec![0; len];
+    input.copy_from_slice(&output);
+    Phase1::deserialize(&output, UseCompression::Yes, CheckForCorrectness::No, &parameters).unwrap();
+
+    let mut output = vec![0; expected_response_length];
+
+    // Construct our keypair using the RNG we created above
+    let current_accumulator_hash = blank_hash();
+    let mut rng = derive_rng_from_seed(&data);
```
Comment (ibaryshnikov) [inline setup1-contributor/src/reliability/cpu.rs:30]: is data used only as a seed?

**Comment by ibaryshnikov** [comment]
Looks good! How do I compose the data? I can see there's `UseCompression::Yes`, does it mean that I have to encode the data in a spectial way while generating it on the coordinator side?

File: setup1-contributor/src/reliability/cpu.rs (hunk 1 above)
Comment (emmorais) [reply setup1-contributor/src/reliability/cpu.rs:78]: The input data is passed to construct the RNG, which is used to generate the key pair. Then we can consider the input data as a seed. Using this seed we obtain non-zero key pair, and the computation of the powers of tau works as usual.

**Review by ibaryshnikov — APPROVED** [review]

File: setup1-contributor/src/reliability/cpu.rs (hunk 2 above)
Comment (emmorais) [reply setup1-contributor/src/reliability/cpu.rs:30]: yes

**Comment by emmorais** [comment]
> Looks good! How do I compose the data? I can see there's `UseCompression::Yes`, does it mean that I have to encode the data in a special way while generating it on the coordinator side?

In summary, the use of compression; or the utilization of Marlin; or the choice for batch size; all those choices were basically arbitrary. I just varied TOTAL_SIZE to get reasonable amount of time to finish the computation. In order to verify the output is correct, we can simply rerun the same function using the same data input. This is faster than verifying the contribution itself.

**Review by apruden2008 — APPROVED** [review]

**Pull request merged**